File: srcs/django_code/transcend/myapp/static/js/game/input_handler.js
import { Vec2 } from './Vec2.js';

export function input_id(core, button, inputs) {
	const set = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
	for (let key in inputs) {
		if (set.includes(key) && button.name.length < 4)
			button.name += key.toUpperCase();
		else if (key == "Backspace")
			button.name = button.name.slice(0, -1);
		else if (key == "Enter")
			core.menu.setValues("JOIN", core);
		delete inputs[key];
	}
}

function online_input(core, inputs) {
	if (core.state === "game") {
		for (let key in inputs) {
			if (core.players[core.id - 1].side === "left" || core.players[core.id - 1].side === "right") {
				if (key == "ArrowUp" || key.toUpperCase() == "W") {
					core.pressed.push("UP");
					core.players[core.id - 1].moveUp(core.walls);
				}
				else if (key == "ArrowDown" || key.toUpperCase() == "S") {
					core.pressed.push("DOWN");
					core.players[core.id - 1].moveDown(core.walls);
				}
			}
			if (core.players[core.id - 1].side === "up" || core.players[core.id - 1].side === "down") {
				if (key == "ArrowLeft" || key.toUpperCase() == "A") {
					core.pressed.push("LEFT");
					core.players[core.id - 1].moveLeft(core.walls);
				}
				else if (key == "ArrowRight" || key.toUpperCase() == "D") {
					core.pressed.push("RIGHT");
					core.players[core.id - 1].moveRight(core.walls);
				}
			}
			if (key == " " && core.ball.stick == core.id) {
				core.pressed.push("LAUNCH");
				core.ball.launch();
			}
		}
	}
	if (core.mode === "solo") {
		core.pressed = [];
		ai_moves(core, core.players[1]);
	}
}

export function input_handler(core, inputs) {
	if (core.online || (core.mode === "solo" && !core.pause[0]))
		online_input(core, inputs);
	else {
		if (core.state === "game" && !core.pause[0]) {
			for (let player of core.players) {
				if (player.name === "AI")
					ai_moves(core, player);
				else
					player_moves(core, player, inputs);
			}
		}
	}
}

export function mouse_handler(core, pos) {
	if (core.state === "end")
		core.end.click(core, pos);
	else if (core.state === "menu")
		core.menu.click(core, pos);
	else if (core.pause[0])
		core.pause[1].click(core, pos);
	else if (core.state === "custom")
		core.custom_menu.click(core, pos);
	else if (core.state === "waiting")
		core.wait_screen.click(core, pos);
}

function ai_moves(core, player) {
	for (let ai of core.ai) {
		if (ai.id === player.nb) {
			for (let move of ai.moves) {
				if (move === "UP")
					player.moveUp(core.walls);
				else if (move === "DOWN")
					player.moveDown(core.walls);
				else if (move === "LEFT")
					player.moveLeft(core.walls);
				else if (move === "RIGHT")
					player.moveRight(core.walls);
				else if (move === "LAUNCH" && core.ball.stick === player.nb)
					core.ball.launch();
			}
			ai.pos = new Vec2(player.paddle[0].pos.x, player.paddle[0].pos.y);
			ai.moves = [];
			break;
		}
	}
}

function player_moves(core, player, inputs) {
	for (let key in inputs) {
		if (player.side === "left") {
			if (player.nb == 1) {
				if (key.toUpperCase() == "W")
					player.moveUp(core.walls);
				if (key.toUpperCase() == "S")
					player.moveDown(core.walls);
			}
			else {
				if (key.toUpperCase() == "T")
					player.moveUp(core.walls);
				if (key.toUpperCase() == "G")
					player.moveDown(core.walls);
			}
			if (key == " " && core.ball.stick == player.nb)
				core.ball.launch();
		}
		else if (player.side === "right") {
			if (player.nb == 2 || player.nb == 4) {
				if (key == "ArrowUp")
					player.moveUp(core.walls);
				if (key == "ArrowDown")
					player.moveDown(core.walls);
			}
			else {
				if (key == "8")
					player.moveUp(core.walls);
				if (key == "5")
					player.moveDown(core.walls);
			}
			if (key == "0" && core.ball.stick == player.nb)
				core.ball.launch();
		}
		else if (player.side === "up") {
			if (key.toUpperCase() == "T")
					player.moveLeft(core.walls);
			if (key.toUpperCase() == "Y")
					player.moveRight(core.walls);
			if (key.toUpperCase() == "H" && core.ball.stick == player.nb)
				core.ball.launch();
		}
		else if (player.side === "down") {
			if (key.toUpperCase() == "K")
				player.moveLeft(core.walls);
			if (key.toUpperCase() == "L")
				player.moveRight(core.walls);
			if (key.toUpperCase() == "O" && core.ball.stick == player.nb)
				core.ball.launch();
		}
	}
}

export function escape_handler(core) {
	if (core.state === "end" || core.state === "custom") {
		core.state = "menu";
		core.mode = "none";
		core.max_score = 10;
	}
	if (core.state === "game") {
		core.pause[0] = !core.pause[0];
		if (!core.online && core.pause[0])
			core.pause[1].freeze = true;
		else
			core.pause[1].freeze = false;
	}
	delete core.inputs["Escape"];
}

File: Game Client/game/input_handler.py
from .config import *
import string
from .Vec2 import *

async def input_id(core, button, key, code):
	set = string.ascii_letters + string.digits
	if code in set and button.name.__len__() < 4:
		button.name += code.upper()
	elif key == pg.K_BACKSPACE:
		button.name = button.name[:-1]
	elif key == pg.K_RETURN or key == pg.K_KP_ENTER:
		await core.menu.setValues("JOIN", core)

def online_input(core):
	if core.state == "game":
		if core.players[core.id - 1].side == "left" or core.players[core.id - 1].side == "right":
			if core.keyboardState[pg.K_UP] or core.keyboardState[pg.K_w]:
				core.pressed.append("UP")
				core.players[core.id - 1].moveUp(core.walls)
			if core.keyboardState[pg.K_DOWN] or core.keyboardState[pg.K_s]:
				core.pressed.append("DOWN")
				core.players[core.id - 1].moveDown(core.walls)
		if core.players[core.id - 1].side == "up" or core.players[core.id - 1].side == "down":
			if core.keyboardState[pg.K_LEFT] or core.keyboardState[pg.K_a]:
				core.pressed.append("LEFT")
				core.players[core.id - 1].moveLeft(core.walls)
			if core.keyboardState[pg.K_RIGHT] or core.keyboardState[pg.K_d]:
				core.pressed.append("RIGHT")
				core.players[core.id - 1].moveRight(core.walls)
		if core.keyboardState[pg.K_SPACE] and core.ball.stick == core.id:
			core.pressed.append("LAUNCH")
			core.ball.launch()
	if core.mode == 'solo':
		core.pressed = []
		ai_moves(core, core.players[1])

def input_handler(core):
	if core.online or (core.mode == "solo" and not core.pause[0]):
		online_input(core)
	else:
		if core.state == "game" and not core.pause[0]:
			for player in core.players:
				if player.name == 'AI':
					ai_moves(core, player)
				else:
					player_moves(core, player)


async def mouse_handler(core):
	if core.state == "end":
		await end_input(core)
	elif core.state == "menu":
		await menu_input(core)
	elif core.pause[0]:
		await pause_input(core)
	elif core.state == "custom":
		await custom_input(core)
	elif core.state == 'waiting':
		if core.mouseState[0] and pg.mouse.get_focused():
			await core.wait_screen.click(core, core.mousePos)

async def custom_input(core):
	if core.mouseState[0] and pg.mouse.get_focused():
		await core.custom_menu.click(core, core.mousePos)

async def end_input(core):
	if core.mouseState[0] and pg.mouse.get_focused():
		await core.end.click(core, core.mousePos)

async def menu_input(core):
	if core.mouseState[0] and pg.mouse.get_focused():
		await core.menu.click(core, core.mousePos)

async def pause_input(core):
	if core.mouseState[0] and pg.mouse.get_focused():
		await core.pause[1].click(core, core.mousePos)


def ai_moves(core, player):
	for ai in core.ai:
		if ai.id == player.nb:
			for move in ai.moves:
				if move == "UP":
					player.moveUp(core.walls)
				elif move == "DOWN":
					player.moveDown(core.walls)
				elif move == "LEFT":
					player.moveLeft(core.walls)
				elif move == "RIGHT":
					player.moveRight(core.walls)
				elif move == "LAUNCH" and core.ball.stick == player.nb:
					core.ball.launch()
			ai.pos = Vec2(pos=player.paddle[0].pos)
			ai.moves = []
			break

def player_moves(core, player):
	if player.side == 'left':
		if player.nb == 1:
			if core.keyboardState[pg.K_w]:
				player.moveUp(core.walls)
			if core.keyboardState[pg.K_s]:
				player.moveDown(core.walls)
		else:
			if core.keyboardState[pg.K_t]:
				player.moveUp(core.walls)
			if core.keyboardState[pg.K_g]:
				player.moveDown(core.walls)
		if core.keyboardState[pg.K_SPACE] and core.ball.stick == player.nb:
			core.ball.launch()
	elif player.side == 'right':
		if player.nb == 2 or player.nb == 4:
			if core.keyboardState[pg.K_UP]:
				player.moveUp(core.walls)
			if core.keyboardState[pg.K_DOWN]:
				player.moveDown(core.walls)
		else:
			if core.keyboardState[pg.K_KP8]:
				player.moveUp(core.walls)
			if core.keyboardState[pg.K_KP5]:
				player.moveDown(core.walls)
		if core.keyboardState[pg.K_KP_0] and core.ball.stick == player.nb:
			core.ball.launch()
	elif player.side == 'up':
		if core.keyboardState[pg.K_t]:
			player.moveLeft(core.walls)
		if core.keyboardState[pg.K_y]:
			player.moveRight(core.walls)
		if core.keyboardState[pg.K_h] and core.ball.stick == player.nb:
			core.ball.launch()
	elif player.side == 'down':
		if core.keyboardState[pg.K_k]:
			player.moveLeft(core.walls)
		if core.keyboardState[pg.K_l]:	
			player.moveRight(core.walls)
		if core.keyboardState[pg.K_o] and core.ball.stick == player.nb:
			core.ball.launch()


async def escape_handler(core):
	if core.state == "menu" or core.state == "end":
		await core.quit()
	if core.state == "game":
		core.pause[0] = not core.pause[0]
		if not core.online and core.pause[0]:
			core.pause[1].freeze = True
		else:
			core.pause[1].freeze = False
	if core.state == "custom":
		core.state = "menu"
		core.mode = "none"
		core.max_score = 10
